analyses/focal-cn-file-preparation/05-define-most-focal-cn-units.Rmd

```diff
@@ -0,0 +1,357 @@
+---
+title: "Find most focal recurrent copy number units"
+output: 
+  html_notebook:
+    toc: TRUE
+    toc_float: TRUE
+author: Chante Bethell for ALSF CCDL
+date: 2020
+params:
+  is_ci: FALSE
+---
+
+This notebook defines the most focal recurrent copy number units by removing focal changes that are within entire chromosome arm losses and gains.
+_Most focal_ here meaning:
+
+- If a chromosome arm is not clearly defined as a gain or loss (and is callable) we look to define the cytoband level status
+- If a cytoband is not clearly defined as a gain or loss (and is callable) we then look to define the gene level status
+
+## Usage
+
+This notebook is intended to be run from the command line with the following (assumes you are in the root directory of the repository):
+
+```
+Rscript -e "rmarkdown::render('analyses/focal-cn-file-preparation/05-define-most-focal-cn-units.Rmd', clean = TRUE)"
+```
+
+## Set up
+
+### Libraries and functions
+
+```{r}
+library(tidyverse)
+```
+
+### Custom Function
+
+```{r}
+status_majority_caller <- function(status_df,
+                                   region_variable) {
+  # Given a data.frame with cytoband/gene level copy number status data,
+  # find the dominant status of the cytoband/gene region by calculating
+  # the percentages of the region that each call represents.
+  #
+  # Args:
+  #   status_df: data.frame with cytoband/gene level copy number status data
+  #   region_variable: string of the column name/region to calculate copy number
+  #                    status percentages for
+  # Return:
+  #   status_count: data.frame with percentage values for each unique status in
+  #                 each unique region (arm/cytoband/gene) and the dominant
+  #                 status for that region
+
+  region_sym <- rlang::sym(region_variable)
+
+  if (region_variable == "chromosome_arm") {
+    status_count <- status_df %>%
+      count(!!region_sym, dominant_status) %>%
+      # Spread the data -- each row represents a unique chromosome arm
+      spread(dominant_status, n) %>%
+      replace_na(list(
+        gain = 0,
+        loss = 0,
+        neutral = 0
+      )) %>%
+      group_by(!!region_sym) %>%
+      # The logic here is to define the arm status based on the majority of calls
+      # in the region -- if the number of calls for a specific status is
+      # responsible for more than 50% of the total calls in that region, then it is
+      # used to define the region's status
+      mutate(
+        percent_gain = gain / (gain + loss + neutral),
+        percent_loss = loss / (gain + loss + neutral),
+        percent_neutral = neutral / (gain + loss + neutral),
+        dominant_status = case_when(
+          percent_gain > 0.5 &
+            percent_gain > 1.5 * percent_loss ~ "gain",
```

Comment:
Can you explain these logic statements? I'm not saying they aren't what we want, but I wanna understand the changes you made here and what you are intending.

Reply:
I am sure that I probably went about implementing this logic is the least efficient way (it's probably not doing what I intended), but the idea here is that we want the status percentage to be responsible for more than 50% of calls in that region _and_ to be greater than 1.5 times the alternate status call.

This was an attempt to account for a loss/gain dominant status call being made when there is just slim difference between calls.

Reply:
Ah. Okay. Well as a more general statement, I think whatever the logic is that we have for deciding majority status we should probably these methods pretty upfront laid out in documentation (probably at the top of the notebook and READMEs as well as comment). Not only what the methods are for the majority status call decisions but also what was our reasoning behind that.

Reply:
Gotcha, will do.

Reply:
That being said, I think we are getting closer, but there are two things that we may want to think about when calling majority status (and these are related). 
1) Not calling a region if too much of it is uncallable
2) Not calling a region if the absolute number of calls are too small (related to 1) e.g. 3 calls vs 2 calls would technically be greater than 50% but we probably don't have enough data to make that call. 

But lastly, I'm wondering whether we need a > 0.5 cutoff in addition to a 1.5x cutoff or if raising the threshold of the percent cutoff might establish the same thing?

Reply:
> But lastly, I'm wondering whether we need a > 0.5 cutoff in addition to a 1.5x cutoff or if raising the threshold of the percent cutoff might establish the same thing?

Good question, I'll test this out and include my findings in the notebook (in addition to the points you made above).

Reply:
No need do an entire side analysis but, my point is just that whatever decision logic we come up with we should make sure we are clear about what it is and why we think this gives us a reasonable answer.